src/core/mkShell.nix

```diff
@@ -0,0 +1,262 @@
+      enable = shell.preCommit.png-optimization.enable;
+      extraOptions = shell.preCommit.png-optimization.extraOptions;
+      package = shell-tools.optipng;
+      include = [ "png" ];
+    };
+
+    purs-tidy = {
+      enable = shell.preCommit.purs-tidy.enable;
+      extraOptions = shell.preCommit.purs-tidy.extraOptions;
+      options = "format-in-place";
+      package = shell-tools.purs-tidy;
+      include = [ "purs" ];
+    };
+  };
+
+
+  mkPreCommitHook = name: hook: lib.mkForce {
+    entry =
+      "${lib.getExe hook.package} " +
+      "${utils.getAttrWithDefault "options" "" hook} " +
+      "${utils.getAttrWithDefault "extraOptions" "" hook}";
+
+    files =
+      if hook ? include
+      then "\\.(${lib.concatStringsSep "|" hook.include})$"
+      else "";
+
+    enable = hook.enable;
+    name = name;
+    pass_filenames = true;
+    types = utils.getAttrWithDefault "types" [ ] hook;
+  };
+
+
+  pre-commit-check = iogx-inputs.pre-commit-hooks-nix.lib.${system}.run {
+    src = lib.cleanSource user-inputs.self;
+    hooks = lib.mapAttrs mkPreCommitHook pre-commit-hooks;
+  };
+
+
+  pre-commit-profile = {
+    packages = [ pkgs.pre-commit ];
+    shellHook = pre-commit-check.shellHook;
+  };
+
+  ### merging ##################################################################
+
+  shell-as-shell-profile =
+    removeAttrs shell [ "name" "prompt" "welcomeMessage" ];
+
+
+  name-and-welcome-message-profile = {
+    shellHook = ''  
+      export PS1="${shell.prompt}"
+      echo 
+      printf "${shell.welcomeMessage}"
+      echo
+      echo
+      echo "Type 'info' to see what's inside this shell."
+    '';
+  };
+
+
+  local-archive-profile.env.LOCALE_ARCHIVE = lib.optionalString
+    (pkgs.stdenv.hostPlatform.libc == "glibc")
+    ("${pkgs.glibcLocales}/lib/locale/locale-archive");
+
+
+  base-profile = repo.src.core.mkMergedShell (extraProfiles ++ [
+    pre-commit-profile
+    shell-as-shell-profile
+    local-archive-profile
+    toolchain-profile
+    name-and-welcome-message-profile
+  ]);
+
+
+  utility-scripts-profile = {
+    scripts = repo.src.core.mkShellUtilityScripts base-profile;
+  };
+
+
+  final-profile = repo.src.core.mkMergedShell [ base-profile utility-scripts-profile ];
+
+
+  final-scripts-as-packages =
+    let
+      removeDisabled = lib.filterAttrs (_: { enable ? true, ... }: enable);
+      enabled-scripts = removeDisabled final-profile.scripts;
+      scriptToPackage = name: script: pkgs.writeShellScriptBin name "${script.exec}";
+    in
+    lib.mapAttrsToList scriptToPackage enabled-scripts;
+
+
+  final-env-as-bash =
+    let exportVar = key: val: ''export ${key}="${toString val}"'';
+    in lib.concatStringsSep "\n" (lib.mapAttrsToList exportVar final-profile.env);
+
+
+  dev-shell = pkgs.mkShell {
+    name = shell.name;
+    buildInputs = final-profile.packages ++ final-scripts-as-packages;
+    shellHook = ''  
+      ${final-profile.shellHook}
+      ${final-env-as-bash}
+    '';
+  };
```

Comment:
@coderabbitai the shellHook ins't really hard-coded. You are getting confused because you see a literal string, but we are doing string interpolation using `${}` syntax here.

Reply:
@coderabbitai I think I'm starting to understand your notion of complexity.
You don't like complicated expressions, and you suggest that they be split into functions when they grow too large.
That's generally true. In this case though `shellHook` is not complex enough to warrant it's own function/variable.